### PR TITLE
Start using Gradle Tooling API

### LINKED ISSUE #3497: Use gradle model instead of string parsing for manipulating gradle files
**Description**
Actually we use string parsing and manipulation to add extension, complete existing gradle file, or simply read gradle files. It's fragil, On maven side, we use an external model and parser. 

**Implementation ideas**
I am just dig a little more to not use string and delegate parse of model to gradle lib.
I see how eclipse do it:
https://github.com/eclipse/buildship/blob/master/org.eclipse.buildship.kotlin/build.gradle
https://github.com/eclipse/buildship/blob/master/org.eclipse.buildship.kotlin/src/main/java/org/eclipse/buildship/kotlin/internal/GradleKotlinScriptTemplateProvider.java#L80

And with ProjectConnection we can load the groovy model too, but I don't know the limitation and we need all gradle core as dependency. 

see https://github.com/quarkusio/quarkus/pull/3475#discussion_r313254631

### PR DESCRIPTION
**Pull request by Dufgui**
fix #3497

### CODE REVIEW

**Comment by Dufgui** [comment]
normally ready for review @geoand @aloubyansky

**Comment by geoand** [comment]
Tests seem to be failing:

```
[ERROR] Tests run: 8, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 2.19 s <<< FAILURE! - in io.quarkus.cli.commands.CreateProjectTest
[ERROR] createGradle  Time elapsed: 1.861 s  <<< ERROR!
org.gradle.tooling.BuildException: Could not fetch model of type 'EclipseProject' using Gradle distribution 'https://services.gradle.org/distributions/gradle-5.5.1-bin.zip'.
```

**Comment by Dufgui** [comment]
I think it's now fixed, the remaining error seems to be an instability from the test job (tomeout) @geoand @aloubyansky

**Comment by gsmet** [comment]
It might be a stupid question but why do we have Gradle specific classes in `devtools/common`?

Not exactly related to your patch but adding a dependency to Gradle + the Gradle Maven repo there seems a bit of a problem to me.

**Comment by Dufgui** [comment]
We already have a dependence on model from maven. And I want to produce the
same feature in gradle and maven.


Le lun. 19 août 2019 à 11:56, Guillaume Smet <notifications@github.com> a
écrit :

**Comment by Dufgui** [comment]
But I can do a refactoring to remove maven and gradle dep from their ?

Le lun. 19 août 2019 à 12:26, Guillaume Dufour <guillaume.duff@gmail.com> a
écrit :

**Comment by Dufgui** [comment]
Or you want to keep maven as core. And overwrite only gradle version ?

**Comment by Dufgui** [comment]
I have the same problem in bootstrap core. With #3552   What do you prefere @gsmet

**Comment by Dufgui** [comment]
so, what could anyone, relaunch the tests and review please ?
@gsmet @aloubyansky @geoand  ?

File: devtools/common/src/test/resources/gradle-project/gradle.properties [hunk 1]
```diff
@@ -1 +1 @@
-quarkusVersion = 999-SNAPSHOOT
\ No newline at end of file
+quarkusVersion = 0.21.1
```
Comment (geoand) [inline devtools/common/src/test/resources/gradle-project/gradle.properties]: This is probably not a good question and I am assuming that I am missing something fundamental here, but how would this work with the current code if the test uses a specific version?

**Comment by Dufgui** [comment]
You are right. I must migrate this test to gradle plugin.

**Comment by Dufgui** [comment]
gradle dep removed from common and put in gradle project.
So the maven mojo cannot list extension version of a gradle project now.

**Comment by Dufgui** [comment]
@aloubyansky it's works !!!!! you can review

File: devtools/gradle/src/test/java/io/quarkus/gradle/GradleBuildFileTest.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+    private static GradleBuildFileFromConnector buildFile;
+
+    @BeforeAll
+    public static void beforeAll() throws IOException, URISyntaxException {
+        URL url = GradleBuildFileTest.class.getClassLoader().getResource("gradle-project");
+        URI uri = url.toURI();
+        Path gradleProjectPath;
+        if (uri.getScheme().equals("jar")) {
+            FileSystem fileSystem = FileSystems.newFileSystem(uri, Collections.<String, Object> emptyMap());
```
Comment (aloubyansky) [inline devtools/gradle/src/test/java/io/quarkus/gradle/GradleBuildFileTest.java]: Normally this should be a directory, right? When would that be found in a JAR? FileSystem should be closed then.

**Review by aloubyansky — COMMENTED** [review]
Well done, @Dufgui! I added a question there.

File: devtools/gradle/src/test/java/io/quarkus/gradle/GradleBuildFileTest.java (hunk 2 above)
Comment (Dufgui) [reply devtools/gradle/src/test/java/io/quarkus/gradle/GradleBuildFileTest.java]: you are right

**Review by aloubyansky — APPROVED** [review]
I'm approving this PR but I also see a few things that need to be fixed as a follow up (they are not introduced by this specific PR). It also doesn't necessarily mean they all fall on your shoulders @Dufgui :) If you like to continue improving this area - great, if not, that's fine too, your contributions are always appreciated.
Here is what I see needs to be improved:
* proper logger (there are `System.out.println` calls in various places, there is also BuildFile.PRINTER, if we can't use jboss-logging directly then we should use BuildFile.PRINTER or another abstraction everywhere);
* we should look into replacing mentions of `999-SNAPSHOT` in gradle build scripts with `${version}` that's set in one place like `gradle.properties`.

Actually the bigger task should be to move the gradle plugin into a separate repo and be a pure Gradle project. I think that's the requirement to be accepted into the Gradle plugin repo.

**Comment by kenfinnigan** [comment]
Is this good to be merged?

**Pull request merged**